Token healing

ahmed-moubtahij opened the pull request:
# What does this PR do?

Token healing rectifies the token boundary bias in greedy tokenization. It does this by trimming and regrowing the prompt to better align with the model's tokenizer, thus enhancing generation quality. The improvement is clearest with completion models.

Token boundary bias is a silent performance killer that doesn't seem very well known. It has clear impact on completion quality, though I'm not sure where it would fit as a transformers feature.

A more thorough explanation of the problem: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38).

### Motivation

Given a completion prompt with a partial url ending with `:`, the model might have seen the expected completion `://` as a _single_ token in training. However, the prompt's tail token `:` tells it that the next token is not `//`, and so it generates a wrong completion. Such errors compound in auto-regressive language models.


Fixes [#28346](https://github.com/huggingface/transformers/issues/28346)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

- @gante

ahmed-moubtahij committed:
first working commit of token healing

ahmed-moubtahij committed:
rebase with main

ahmed-moubtahij committed:
first working commit of token healing

ahmed-moubtahij committed:
first working commit of token healing

ahmed-moubtahij committed:
first working commit of token healing

ahmed-moubtahij committed:
first working commit of token healing

ahmed-moubtahij committed:
update doc

ahmed-moubtahij committed:
update doc

ahmed-moubtahij committed:
update doc

ahmed-moubtahij committed:
update doc

ahmed-moubtahij committed:
batched generation + isolated token healing

ahmed-moubtahij committed:
batched generation + isolated token healing

ahmed-moubtahij committed:
batched generation + isolated token healing

ahmed-moubtahij committed:
batched generation + isolated token healing

ahmed-moubtahij committed:
fixed predicted + test-appropriate generation config

ahmed-moubtahij committed:
fixed predicted + test-appropriate generation config

ahmed-moubtahij committed:
fixed predicted + test-appropriate generation config

ahmed-moubtahij committed:
fixed predicted + test-appropriate generation config

ahmed-moubtahij committed:
token healing impl

ahmed-moubtahij committed:
token healing impl

ahmed-moubtahij committed:
token healing impl

ahmed-moubtahij committed:
token healing impl

ahmed-moubtahij committed:
token healing impl

ahmed-moubtahij committed:
token healing impl

ahmed-moubtahij committed:
tests pass

ahmed-moubtahij committed:
tests pass

ahmed-moubtahij committed:
tests pass

ahmed-moubtahij committed:
tests pass

ahmed-moubtahij committed:
tests pass

ahmed-moubtahij committed:
applied make fixup

ahmed-moubtahij committed:
applied make fixup

ahmed-moubtahij committed:
applied make fixup

ahmed-moubtahij committed:
applied make fixup

ahmed-moubtahij committed:
applied make fixup

ahmed-moubtahij committed:
Merge branch 'token_healing' of https://github.com/Ayenem/transformers into token_healing

ahmed-moubtahij committed:
docstring + correctly checking for absence of token alternatives

ahmed-moubtahij committed:
docstring + correctly checking for absence of token alternatives

ahmed-moubtahij committed:
docstring + correctly checking for absence of token alternatives

Ayenem commented:
Circleci tests seem to fail, `tests_flax` for example outputs:

> ImportError while importing test module '/home/circleci/transformers/tests/generation/test_token_healing.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> ../.pyenv/versions/3.8.12/lib/python3.8/importlib/__init__.py:127: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> tests/generation/test_token_healing.py:23: in <module>
>     class TokenHealingTestCase(unittest.TestCase):
> tests/generation/test_token_healing.py:26: in TokenHealingTestCase
>     completion_model = AutoModelForCausalLM.from_pretrained(
> ../.pyenv/versions/3.8.12/lib/python3.8/site-packages/transformers/utils/import_utils.py:1304: in __getattribute__
>     requires_backends(cls, cls._backends)
> ../.pyenv/versions/3.8.12/lib/python3.8/site-packages/transformers/utils/import_utils.py:1292: in requires_backends
>     raise ImportError("".join(failed))
> E   ImportError: 
> E   AutoModelForCausalLM requires the PyTorch library but it was not found in your environment.

I don't know how to specify to the Circleci environment that it should install pytorch. This is my first PR, so I would appreciate some guidance 🙏

gante commented on examples/pytorch/token-healing/README.md:
```diff
@@ -0,0 +1,44 @@
+<!-- back to top link -->
```
`examples/research/token-healing` would be a better place to hold this example :) The examples in `examples/pytorch` are selected by the `transformers` team and actively maintained 🤗

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1168,6 +1168,7 @@ def generate(
         self,
         inputs: Optional[torch.Tensor] = None,
         generation_config: Optional[GenerationConfig] = None,
+        token_healing: bool = False,
```
This new flag should go [here](https://github.com/huggingface/transformers/blob/de13a951b38b85195984164819f1ab05fe508677/src/transformers/generation/configuration_utils.py#L260) instead, with a corresponding entry in the docstring above.

By placing it in the place I'm suggesting, users can still do `model.generate(..., token_healing=true)`, but it has the following advantages:
1. It's consistent with how we handle options in `generate`
2. Doesn't increase the signature of `generate`
3. Owners of models can save `token_healing=True` in their `generation_config`, so that their users can benefit from it by default if the owner believes it is relevant for the model 🙌

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1705,6 +1709,69 @@ def typeerror():
+            `torch.LongTensor` where each sequence has its tail token replaced with its appropriate extension.
+        """
+
+        from pygtrie import CharTrie
+
+        from ..models.auto import AutoTokenizer
```
These imports should go to the top. It also means `pygtrie` should be added to  [this list](https://github.com/huggingface/transformers/blob/de13a951b38b85195984164819f1ab05fe508677/setup.py#L97) ⚠️

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1705,6 +1709,69 @@ def typeerror():
                 **model_kwargs,
             )
 
+    @torch.no_grad()
```
This one is not needed, as `generate` already has it

```suggestion
```

gante commented on tests/generation/test_token_healing.py:
```diff
@@ -0,0 +1,55 @@
+
+from parameterized import parameterized
+
+from transformers import AutoModelForCausalLM, AutoTokenizer
```
This import is what's breaking CI: the test fetcher for all CI runs has to execute this line and, for test images without pytorch installed, this line will fail :)

Read my comment below for a suggested fix

gante commented on tests/generation/test_token_healing.py:
```diff
@@ -0,0 +1,55 @@
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+
+class TokenHealingTestCase(unittest.TestCase):
```
I'm going to ask to move this test class to `tests/generation/test_utils.py`. By default, we only create a new file when there is significant content for the file in question 🤗 

The imports on the file I'm suggesting are correct, so it remove the CI issues!

gante reviewed (commented):
Thank you for opening this cool PR! I have written a few comments, but they should be easy to solve 🤗

ahmed-moubtahij committed:
conforming file structure + token_healing in generation_config + fixed imports

ahmed-moubtahij committed:
conforming file structure + token_healing in generation_config + fixed imports

ahmed-moubtahij committed:
conforming file structure + token_healing in generation_config + fixed imports

Ayenem commented:
> Thank you for opening this cool PR! I have written a few comments, but they should be easy to solve 🤗

Thank you for your detailed reviews and for taking the time to explain their reasons 🙏 Hopefully, they're all fixed now.

ahmed-moubtahij committed:
Merge branch 'main' into token_healing

poedator committed:
`HfQuantizer` class for quantization-related stuff in `modeling_utils.py` (#26610)

* squashed earlier commits for easier rebase

* rm rebase leftovers

* 4bit save enabled @quantizers

* TMP gptq test use exllama

* fix AwqConfigTest::test_wrong_backend for A100

* quantizers AWQ fixes

* _load_pretrained_model low_cpu_mem_usage branch

* quantizers style

* remove require_low_cpu_mem_usage attr

* rm dtype arg from process_model_before_weight_loading

* rm config_origin from Q-config

* rm inspect from q_config

* fixed docstrings in QuantizationConfigParser

* logger.warning fix

* mv is_loaded_in_4(8)bit to BnbHFQuantizer

* is_accelerate_available error msg fix in quantizer

* split is_model_trainable in bnb quantizer class

* rm llm_int8_skip_modules as separate var in Q

* Q rm todo

* fwd ref to HFQuantizer in type hint

* rm note re optimum.gptq.GPTQQuantizer

* quantization_config in __init__ simplified

* replaced NonImplemented with  create_quantized_param

* rm load_in_4/8_bit deprecation warning

* QuantizationConfigParser refactoring

* awq-related minor changes

* awq-related changes

* awq config.modules_to_not_convert

* raise error if no q-method in q-config in args

* minor cleanup

* awq quantizer docstring

* combine common parts in bnb process_model_before_weight_loading

* revert test_gptq

* .process_model_ cleanup

* restore dict config warning

* removed typevars in quantizers.py

* cleanup post-rebase 16 jan

* QuantizationConfigParser classmethod refactor

* rework of handling of unexpected aux elements of bnb weights

* moved q-related stuff from save_pretrained to quantizers

* refactor v1

* more changes

* fix some tests

* remove it from main init

* ooops

* Apply suggestions from code review

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

* fix awq issues

* fix

* fix

* fix

* fix

* fix

* fix

* add docs

* Apply suggestions from code review

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>
Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Update docs/source/en/hf_quantizer.md

* address comments

* fix

* fixup

* Update src/transformers/modeling_utils.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Update src/transformers/modeling_utils.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* address final comment

* update

* Update src/transformers/quantizers/base.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Update src/transformers/quantizers/auto.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* fix

* add kwargs update

* fixup

* add `optimum_quantizer` attribute

* oops

* rm unneeded file

* fix doctests

---------

Co-authored-by: younesbelkada <younesbelkada@gmail.com>
Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>
Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>
Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>
Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

ahmed-moubtahij committed:
make fixup

ahmed-moubtahij committed:
make fixup

younesbelkada committed:
[`HfQuantizer`] Move it to "Developper guides" (#28768)

Update _toctree.yml

gau-nernst committed:
Use Conv1d for TDNN (#25728)

* use conv for tdnn

* run make fixup

* update TDNN

* add PEFT LoRA check

* propagate tdnn warnings to others

* add missing imports

* update TDNN in wav2vec2_bert

* add missing imports

fxmarty committed:
Fix transformers.utils.fx compatibility with torch<2.0 (#28774)

guard sdpa on torch>=2.0

ydshieh committed:
Further pin pytest version (in a temporary way) (#28780)

fix

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

amyeroberts committed:
[`Backbone`] Use `load_backbone` instead of `AutoBackbone.from_config` (#28661)

* Enable instantiating model with pretrained backbone weights

* Remove doc updates until changes made in modeling code

* Use load_backbone instead

* Add use_timm_backbone to the model configs

* Add missing imports and arguments

* Update docstrings

* Make sure test is properly configured

* Include recent DPT updates

amyeroberts committed:
Task-specific pipeline init args (#28439)

* Abstract out pipeline init args

* Address PR comments

* Reword

* BC PIPELINE_INIT_ARGS

* Remove old arguments

* Small fix

Rocketknight1 committed:
Add tf_keras imports to prepare for Keras 3 (#28588)

* Port core files + ESM (because ESM code is odd)

* Search-replace in modelling code

* Fix up transfo_xl as well

* Fix other core files + tests (still need to add correct import to tests)

* Fix cookiecutter

* make fixup, fix imports in some more core files

* Auto-add imports to tests

* Cleanup, add imports to sagemaker tests

* Use correct exception for importing tf_keras

* Fixes in modeling_tf_utils

* make fixup

* Correct version parsing code

* Ensure the pipeline tests correctly revert to float32 after each test

* Ensure the pipeline tests correctly revert to float32 after each test

* More tf.keras -> keras

* Add dtype cast

* Better imports of tf_keras

* Add a cast for tf.assign, just in case

* Fix callback imports

Rocketknight1 committed:
Pin Torch to <2.2.0 (#28785)

* Pin torch to <2.2.0

* Pin torchvision and torchaudio as well

* Playing around with versions to see if this helps

* twiddle something to restart the CI

* twiddle it back

* Try changing the natten version

* make fixup

* Revert "Try changing the natten version"

This reverts commit de0d6592c35dc39ae8b5a616c27285db28262d06.

* make fixup

* fix fix fix

* fix fix fix

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

younesbelkada committed:
[`bnb`] Fix bnb slow tests (#28788)

fix bnb slow tests

codiceSpaghetti committed:
Prevent MLflow exception from disrupting training (#28779)

Modified MLflow logging metrics from synchronous to asynchronous

Co-authored-by: codiceSpaghetti <alessio.ser@hotmail.it>

tom-p-reichel committed:
don't initialize the output embeddings if we're going to tie them to input embeddings (#28192)

* test that tied output embeddings aren't initialized on load

* don't initialize the output embeddings if we're going to tie them to the input embeddings

ahmed-moubtahij committed:
add pygtrie to test requirements

ahmed-moubtahij committed:
add pygtrie to test requirements

ahmed-moubtahij committed:
add pygtrie to test requirements

younesbelkada committed:
[`HFQuantizer`] Remove `check_packages_compatibility` logic (#28789)

remove `check_packages_compatibility` logic

patrickvonplaten committed:
[Whisper] Refactor forced_decoder_ids & prompt ids (#28687)

* up

* Fix more

* Correct more

* Fix more tests

* fix fast tests

* Fix more

* fix more

* push all files

* finish all

* make style

* Fix timestamp wrap

* make style

* make style

* up

* up

* up

* Fix lang detection behavior

* Fix lang detection behavior

* Add lang detection test

* Fix lang detection behavior

* make style

* Update src/transformers/models/whisper/generation_whisper.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

* better error message

* make style tests

* add warning

---------

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

lh0x00 committed:
Resolve DeepSpeed cannot resume training with PeftModel (#28746)

* fix: resolve deepspeed resume peft model issues

* chore: update something

* chore: update model instance pass into is peft model checks

* chore: remove hard code value to tests

* fix: format code

julien-c committed:
canonical repos moves (#28795)

* canonical repos moves

* Style

---------

Co-authored-by: Lysandre <lysandre@huggingface.co>

Rocketknight1 committed:
Wrap Keras methods to support BatchEncoding (#28734)

* Shim the Keras methods to support BatchEncoding

* Extract everything to a convert_batch_encoding function

* Convert BatchFeature too (thanks Amy)

* tf.keras -> keras

kiansierra committed:
Flax mistral (#26943)

* direct copy from llama work

* mistral modules forward pass working

* flax mistral forward pass with sliding window

* added tests

* added layer collection approach

* Revert "added layer collection approach"

This reverts commit 0e2905bf2236ec323163fc1a9f0c016b21aa8b8f.

* Revert "Revert "added layer collection approach""

This reverts commit fb17b6187ac5d16da7c461e1130514dc3d137a43.

* fixed attention outputs

* added mistral to init and auto

* fixed import name

* fixed layernorm weight dtype

* freeze initialized weights

* make sure conversion consideres bfloat16

* added backend

* added docstrings

* added cache

* fixed sliding window causal mask

* passes cache tests

* passed all tests

* applied make style

* removed commented out code

* applied fix-copies ignored other model changes

* applied make fix-copies

* removed unused functions

* passed generation integration test

* slow tests pass

* fixed slow tests

* changed default dtype from jax.numpy.float32 to float32 for docstring check

* skip cache test  for FlaxMistralForSequenceClassification since if pad_token_id in input_ids it doesn't score previous input_ids

* updated checkpoint since from_pt not included

* applied black style

* removed unused args

* Applied styling and fixup

* changed checkpoint for doc back

* fixed rf after adding it to hf hub

* Add dummy ckpt

* applied styling

* added tokenizer to new ckpt

* fixed slice format

* fix init and slice

* changed ref for placeholder TODO

* added copies from Llama

* applied styling

* applied fix-copies

* fixed docs

* update weight dtype reconversion for sharded weights

* removed Nullable input ids

* Removed unnecessary output attentions in Module

* added embedding weight initialziation

* removed unused past_key_values

* fixed deterministic

* Fixed RMS Norm and added copied from

* removed input_embeds

* applied make style

* removed nullable input ids from sequence classification model

* added copied from GPTJ

* added copied from Llama on FlaxMistralDecoderLayer

* added copied from to FlaxMistralPreTrainedModel methods

* fix test deprecation warning

* freeze gpt neox random_params and fix copies

* applied make style

* fixed doc issue

* skipped docstring test to allign # copied from

* applied make style

* removed FlaxMistralForSequenceClassification

* removed unused padding_idx

* removed more sequence classification

* removed sequence classification

* applied styling and consistency

* added copied from in tests

* removed sequence classification test logic

* applied styling

* applied make style

* removed freeze and fixed copies

* undo test change

* changed repeat_kv to tile

* fixed to key value groups

* updated copyright year

* split casual_mask

* empty to rerun failed pt_flax_equivalence test FlaxWav2Vec2ModelTest

* went back to 2023 for tests_pr_documentation_tests

* went back to 2024

* changed tile to repeat

* applied make style

* empty for retry on Wav2Vec2

gante committed:
DeepSpeed: hardcode `torch.arange` dtype on `float` usage to avoid incorrect initialization (#28760)

ydshieh committed:
Add artifact name in job step to maintain job / artifact correspondence (#28682)

* avoid using job name

* apply to other files

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
Split daily CI using 2 level matrix (#28773)

* update / add new workflow files

* Add comment

* Use env.NUM_SLICES

* use scripts

* use scripts

* use scripts

* Fix

* using one script

* Fix

* remove unused file

* update

* fail-fast: false

* remove unused file

* fix

* fix

* use matrix

* inputs

* style

* update

* fix

* fix

* no model name

* add doc

* allow args

* style

* pass argument

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

Ki-Seki committed:
[docs] Correct the statement in the docstirng of compute_transition_scores in generation/utils.py (#28786)

hackyon committed:
Adding [T5/MT5/UMT5]ForTokenClassification (#28443)

* Adding [T5/MT5/UMT5]ForTokenClassification

* Add auto mappings for T5ForTokenClassification and variants

* Adding ForTokenClassification to the list of models

* Adding attention_mask param to the T5ForTokenClassification test

* Remove outdated comment in test

* Adding EncoderOnly and Token Classification tests for MT5 and UMT5

* Fix typo in umt5 string

* Add tests for all the existing MT5 models

* Fix wrong comment in dependency_versions_table

* Reverting change to common test for _keys_to_ignore_on_load_missing

The test is correctly picking up redundant keys in _keys_to_ignore_on_load_missing.

* Removing _keys_to_ignore_on_missing from MT5 since the key is not used in the model

* Add fix-copies to MT5ModelTest

ydshieh committed:
Make `is_torch_bf16_available_on_device` more strict (#28796)

fix

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

fxmarty committed:
Fix symbolic_trace with kv cache (#28724)

* fix symbolic_trace with kv cache

* comment & better test

Rocketknight1 committed:
Add tip on setting tokenizer attributes (#28764)

* Add tip on setting tokenizer attributes

* Grammar

* Remove the bit that was causing doc builds to fail

SangbumChoi committed:
enable graident checkpointing in DetaObjectDetection and add tests in Swin/Donut_Swin (#28615)

* enable graident checkpointing in DetaObjectDetection

* fix missing part in original DETA

* make style

* make fix-copies

* Revert "make fix-copies"

This reverts commit 4041c86c29248f1673e8173b677c20b5a4511358.

* remove fix-copies of DetaDecoder

* enable swin gradient checkpointing

* fix gradient checkpointing in donut_swin

* add tests for deta/swin/donut

* Revert "fix gradient checkpointing in donut_swin"

This reverts commit 1cf345e34d3cc0e09eb800d9895805b1dd9b474d.

* change supports_gradient_checkpointing pipeline to PreTrainedModel

* Revert "add tests for deta/swin/donut"

This reverts commit 6056ffbb1eddc3cb3a99e4ebb231ae3edf295f5b.

* Revert "Revert "fix gradient checkpointing in donut_swin""

This reverts commit 24e25d0a14891241de58a0d86f817d0b5d2a341f.

* Simple revert

* enable deformable detr gradient checkpointing

* add gradient in encoder

zspo committed:
[docs] fix some bugs about parameter description (#28806)

Co-authored-by: p_spozzhang <p_spozzhang@tencent.com>

rajveer43 committed:
Add models from deit (#28302)

* Add modelss

* Add 2 more models

* add models to tocrree

* Add modles

* Update docs/source/ja/model_doc/detr.md

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/ja/model_doc/deit.md

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/ja/model_doc/deplot.md

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* fix bugs

---------

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

stevhliu committed:
[docs] Backbone (#28739)

* backbones

* fix path

* fix paths

* fix code snippet

* fix links

stevhliu committed:
[docs] HfQuantizer (#28820)

* tidy

* fix path

khipp committed:
[Docs] Fix spelling and grammar mistakes (#28825)

* Fix typos and grammar mistakes in docs and examples

* Fix typos in docstrings and comments

* Fix spelling of `tokenizer` in model tests

* Remove erroneous spaces in decorators

* Remove extra spaces in Markdown link texts

skumar951 committed:
Explicitly check if token ID's are None in TFBertTokenizer constructor (#28824)

Add an explicit none-check, since token ids can be 0

jganitkevitch committed:
Add missing None check for hf_quantizer (#28804)

* Add missing None check for hf_quantizer

* Add test, fix logic.

* make style

* Switch test model to Mistral

* Comment

* Update tests/test_modeling_utils.py

---------

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

ydshieh committed:
Fix issues caused by natten (#28834)

try

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
fix / skip (for now) some tests before switch to torch 2.2 (#28838)

* fix / skip some tests before we can switch to torch 2.2

* style

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
Use `-v` for `pytest` on CircleCI  (#28840)

use -v in pytest

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

pacman100 committed:
Reduce GPU memory usage when using FSDP+PEFT (#28830)

support FSDP+PEFT

amyeroberts committed:
Mark `test_encoder_decoder_model_generate` for `vision_encoder_deocder` as flaky (#28842)

Mark test as flaky

dependabot[bot] committed:
Bump dash from 2.3.0 to 2.15.0 in /examples/research_projects/decision_transformer (#28845)

Bump dash in /examples/research_projects/decision_transformer

Bumps [dash](https://github.com/plotly/dash) from 2.3.0 to 2.15.0.
- [Release notes](https://github.com/plotly/dash/releases)
- [Changelog](https://github.com/plotly/dash/blob/dev/CHANGELOG.md)
- [Commits](https://github.com/plotly/dash/compare/v2.3.0...v2.15.0)

---
updated-dependencies:
- dependency-name: dash
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

VeryLazyBoy committed:
Support custom scheduler in deepspeed training (#26831)

Reuse trainer.create_scheduler to create scheduler for deepspeed

chenzizhao committed:
[Docs] Fix bad doc: replace save with logging (#28855)

Fix bad doc: replace save with logging

w4ffl35 committed:
Ability to override clean_code_for_run (#28783)

* Add clean_code_for_run function

* Call clean_code_for_run from agent method

Narsil committed:
[WIP] Hard error when ignoring tensors. (#27484)

* [WIP] Hard error when ignoring tensors.

* Better selection/error when saving a checkpoint.

- Find all names we should normally drop (those are in the transformers
  config)
- Find all disjoint tensors (for those we can safely trigger a copy to
  get rid of the sharing before saving)
- Clone those disjoint tensors getting rid of the issue
- Find all identical names (those should be declared in the config
  but we try to find them all anyway.)
- For all identical names:
  - If they are in the config, just ignore them everything is fine
  - If they are not, warn about them.
- For all remainder tensors which are shared yet neither identical NOR
  disjoint. raise a hard error.

* Adding a failing test on `main` that passes here.

* We don't need to keep the subfolder logic in this test.

* Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

---------

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

ArthurZucker committed:
[`Doc`] update contribution guidelines (#28858)

update guidelines

ylacombe committed:
Correct wav2vec2-bert inputs_to_logits_ratio (#28821)

* Correct wav2vec2-bert inputs_to_logits_ratio

* correct ratio

* correct ratio, clean asr pipeline

* refactor on one line

amyeroberts committed:
Image Feature Extraction pipeline (#28216)

* Draft pipeline

* Fixup

* Fix docstrings

* Update doctest

* Update pipeline_model_mapping

* Update docstring

* Update tests

* Update src/transformers/pipelines/image_feature_extraction.py

Co-authored-by: Omar Sanseviero <osanseviero@gmail.com>

* Fix docstrings - review comments

* Remove pipeline mapping for composite vision models

* Add to pipeline tests

* Remove for flava (multimodal)

* safe pil import

* Add requirements for pipeline run

* Account for super slow efficientnet

* Review comments

* Fix tests

* Swap order of kwargs

* Use build_pipeline_init_args

* Add back FE pipeline for Vilt

* Include image_processor_kwargs in docstring

* Mark test as flaky

* Update TODO

* Update tests/pipelines/test_pipelines_image_feature_extraction.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Add license header

---------

Co-authored-by: Omar Sanseviero <osanseviero@gmail.com>
Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

eugen-ajechiloae-clearml committed:
ClearMLCallback enhancements: support multiple runs and handle logging better (#28559)

* add clearml tracker

* support multiple train runs

* remove bad code

* add UI entries for config/hparams overrides

* handle models in different tasks

* run ruff format

* tidy code based on code review

---------

Co-authored-by: Eugen Ajechiloae <eugenajechiloae@gmail.com>

xkszltl committed:
Do not use mtime for checkpoint rotation. (#28862)

Resolve https://github.com/huggingface/transformers/issues/26961

nakranivaibhav committed:
Adds LlamaForQuestionAnswering class in modeling_llama.py along with AutoModel Support  (#28777)

* This is a test commit

* testing commit

* final commit with some changes

* Removed copy statement

* Fixed formatting issues

* Fixed error added past_key_values in the forward method

* Fixed a trailing whitespace. Damn the formatting rules are strict

* Added the copy statement

dependabot[bot] committed:
Bump cryptography from 41.0.2 to 42.0.0 in /examples/research_projects/decision_transformer (#28879)

Bump cryptography in /examples/research_projects/decision_transformer

Bumps [cryptography](https://github.com/pyca/cryptography) from 41.0.2 to 42.0.0.
- [Changelog](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pyca/cryptography/compare/41.0.2...42.0.0)

---
updated-dependencies:
- dependency-name: cryptography
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

khipp committed:
[Docs] Update project names and links in awesome-transformers (#28878)

Update project names and repository links in awesome-transformers

eranhirs committed:
Fix LongT5ForConditionalGeneration initialization of lm_head (#28873)

pacman100 committed:
Raise error when using `save_only_model` with `load_best_model_at_end` for DeepSpeed/FSDP (#28866)

* Raise error when using `save_only_model` with `load_best_model_at_end` for DeepSpeed/FSDP

* Update trainer.py

ydshieh committed:
Fix `FastSpeech2ConformerModelTest` and skip it on CPU (#28888)

* fix

* fix

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
Revert "[WIP] Hard error when ignoring tensors." (#28898)

Revert "[WIP] Hard error when ignoring tensors. (#27484)"

This reverts commit 2da28c4b41bba23969a8afe97c3dfdcbc47a57dc.

ydshieh committed:
unpin torch (#28892)

* unpin torch

* check

* check

* check

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

Wauplin committed:
Explicit server error on gated model (#28894)

khipp committed:
[Docs] Fix backticks in inline code and documentation links (#28875)

Fix backticks in code blocks and documentation links

ydshieh committed:
Hotfix - make `torchaudio` get the correct version in `torch_and_flax_job` (#28899)

* check

* check

* check

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

khipp committed:
[Docs] Add missing language options and fix broken links (#28852)

* Add missing entries to the language selector

* Add links to the Colab and AWS Studio notebooks for ONNX

* Use anchor links in CONTRIBUTING.md

* Fix broken hyperlinks due to spaces

* Fix links to OpenAI research articles

* Remove confusing footnote symbols from author names, as they are also considered invalid markup

Sai-Suraj-27 committed:
fix: Fixed the documentation for `logging_first_step` by removing "evaluate" (#28884)

Fixed the documentation for logging_first_step by removing evaluate.

pacman100 committed:
fix Starcoder FA2 implementation (#28891)

Rocketknight1 committed:
Fix Keras scheduler import so it works for older versions of Keras (#28895)

Fix our schedule import so it works for older versions of Keras

danielkorat committed:
⚠️ Raise `Exception` when trying to generate 0 tokens ⚠️ (#28621)

* change warning to exception

* Update src/transformers/generation/utils.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* validate `max_new_tokens` > 0 in `GenerationConfig`

* fix truncation test parameterization in `TextGenerationPipelineTests`

---------

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ydshieh committed:
Update the cache number (#28905)

* fix

* fix

* fix

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

gante commented on src/transformers/generation/utils.py:
```diff
@@ -33,6 +34,7 @@
     MODEL_FOR_SEQ_TO_SEQ_CAUSAL_LM_MAPPING,
     MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING,
     MODEL_FOR_VISION_2_SEQ_MAPPING,
+    AutoTokenizer,
```
This one should be `from ..models.auto.tokenization_auto import AutoTokenizer` instead

ji-huazhong committed:
Add npu device for pipeline (#28885)

add npu device for pipeline

Co-authored-by: unit_test <test@unit.com>

gante reviewed (approved):
Thank you for iterating 🤗 

To fix the dependency issue, I believe you need to add `pygtrie` to `extras["torch"]` in `setup.py` (L262). Otherwise, `pygtrie` is not installed -- being in the list where you added the dependency is not enough to install the dependency :)

khipp committed:
[Docs] Fix placement of tilde character (#28913)

Fix placement of tilde character

khipp committed:
[Docs] Revert translation of '@slow' decorator (#28912)

SystemPanic committed:
Fix utf-8 yaml load for marian conversion to pytorch in Windows (#28618)

Fix utf-8 yaml in marian conversion

ArthurZucker committed:
[`Core generation`] Adds support for static KV cache (#27931)

Co-authored-by: fxmarty <9808326+fxmarty@users.noreply.github.com>
Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>
Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

Rocketknight1 committed:
Remove dead TF loading code (#28926)

Remove dead code

vodkaslime committed:
fix: torch.int32 instead of torch.torch.int32 (#28883)

zucchini-nlp committed:
pass kwargs in stopping criteria list (#28927)

zucchini-nlp committed:
Support batched input for decoder start ids (#28887)

* support batched input for decoder start ids

* Fix typos

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* minor changes

* fix: decoder_start_id as list

* empty commit

* empty commit

* empty commit

* empty commit

* empty commit

* empty commit

* empty commit

* empty commit

* empty commit

---------

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

khipp committed:
[Docs] Fix broken links and syntax issues (#28918)

* Fix model documentation links in attention.md

* Fix external link syntax

* Fix target anchor names of section links

* Fix copyright statement comments

* Fix documentation headings

ahmed-moubtahij committed:
Update examples/flax/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update examples/flax/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update examples/flax/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update examples/pytorch/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update examples/pytorch/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update examples/pytorch/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update examples/tensorflow/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update examples/tensorflow/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update examples/tensorflow/_tests_requirements.txt

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
fix pygtrie dep

ahmed-moubtahij committed:
fix pygtrie dep

ahmed-moubtahij committed:
fix pygtrie dep

Ayenem commented:
> Thank you for iterating 🤗
> 
> To fix the dependency issue, I believe you need to add `pygtrie` to `extras["torch"]` in `setup.py` (L262). Otherwise, `pygtrie` is not installed -- being in the list where you added the dependency is not enough to install the dependency :)

Thank you for the follow-up! I applied your suggestions. CI is still breaking on the pygtrie import unfortunately. I'll try to think about it some more this weekend.

karl-hajjar committed:
Fix max_position_embeddings default value for llama2 to 4096 #28241 (#28754)

* Changed max_position_embeddings default value from 2048 to 4096

* force push

* Fixed formatting issues. Fixed missing argument in write_model.

* Reverted to the default value 2048 in the Llama config. Added comments for the llama_version argument.

* Fixed issue with default value value of max_position_embeddings in docstring

* Updated help message for llama versions

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

B-Step62 committed:
Fix a wrong link to CONTRIBUTING.md section in PR template (#28941)

peblair committed:
Fix type annotations on neftune_noise_alpha and fsdp_config TrainingArguments parameters (#28942)

khipp committed:
[i18n-de] Translate README.md to German (#28933)

* Translate README.md to German

* Add links to README_de.md

* Remove invisible characters in README

* Change to a formal tone and fix punctuation marks

NielsRogge committed:
[Nougat] Fix pipeline (#28242)

* Fix pipeline

* Remove print statements

* Address comments

* Address issue

* Remove unused imports

NielsRogge committed:
[Docs] Update README and default pipelines (#28864)

* Update README and docs

* Update README

* Update README

KossaiSbai committed:
Convert `torch_dtype` as `str` to actual torch data type (i.e. "float16" …to `torch.float16`) (#28208)

* Convert torch_dtype as str to actual torch data type (i.e. "float16" to torch.float16)

* Check if passed torch_dtype is an attribute in torch

* Update src/transformers/pipelines/__init__.py

Check type via isinstance

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

cmahmut committed:
[`pipelines`] updated docstring with vqa alias (#28951)

updated docstring with vqa alias

gante committed:
Tests: tag `test_save_load_fast_init_from_base` as flaky (#28930)

alekseyfa committed:
Updated requirements for image-classification samples: datasets>=2.14.0 (#28974)

Updated datasets requirements. Need a package version >= 2.14.0

hackyon committed:
Always initialize tied output_embeddings if it has a bias term (#28947)

Continue to initialize tied output_embeddings if it has a bias term

The bias term is not tied, and so will need to be initialized accordingly.

woshiyyya committed:
Clean up staging tmp checkpoint directory (#28848)

clean up remaining tmp checkpoint dir

Signed-off-by: woshiyyya <xiaoyunxuan1998@gmail.com>

khipp committed:
[Docs] Add language identifiers to fenced code blocks (#28955)

Add language identifiers to code blocks

NielsRogge committed:
[Docs] Add video section (#28958)

Add video section

khipp committed:
[i18n-de] Translate CONTRIBUTING.md to German (#28954)

* Translate contributing.md to German

* Fix formatting issues in contributing.md

* Address review comments

* Fix capitalization

ArthurZucker committed:
[`NllbTokenizer`] refactor with added tokens decoder (#27717)

* refactor with addedtokens decoder

* style

* get rid of lang code to id

* style

* keep some things for BC

* update tests

* add the mask token at the end of the vocab

* nits

* nits

* fix final tests

* style

* nits

* Update src/transformers/models/nllb/tokenization_nllb_fast.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

* nits

* style?

* Update src/transformers/convert_slow_tokenizer.py

* make it a tad bit more custom

* ruff please stop
Co-Authored by avidale

<dale.david@mail.ru>

* Update
Co-authored-by: avidale
<dale.david@mail.ru>

* Update
Co-authored-by: avidale <dale.david@mail.ru>

* oupts

* ouft

* nites

* test

* fix the remaining failing tests

* style

* fix failing test

* ficx other test

* temp dir + test the raw init

* update test

* style

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

hiroshi-matsuda-rit committed:
Add sudachi_projection option to BertJapaneseTokenizer (#28503)

* add sudachi_projection option

* Upgrade sudachipy>=0.6.8

* add a test case for sudachi_projection

* Compatible with older versions of SudachiPy

* make fixup

* make style

* error message for unidic download

* revert jumanpp test cases

* format options for sudachi_projection

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* format options for sudachi_split_mode and sudachi_dict_type

* comment

* add tests for full_tokenizer kwargs

* pass projection arg directly

* require_sudachi_projection

* make style

* revert upgrade sudachipy

* check is_sudachi_projection_available()

* revert dependency_version_table and bugfix

* style format

* simply raise ImportError

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* simply raise ImportError

---------

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

gante committed:
Static Cache: load models with MQA or GQA (#28975)

AdityaKane2001 committed:
Update configuration_llama.py: fixed broken link (#28946)

* Update configuration_llama.py: fix broken link

* [Nit] Explicit redirection not required

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

amyeroberts committed:
[`DETR`] Update the processing to adapt masks & bboxes to reflect padding (#28363)

* Update the processing so bbox coords are adjusted for padding

* Just pad masks

* Tidy up, add tests

* Better tests

* Fix yolos and mark as slow for pycocotols

* Fix yolos - return_tensors

* Clarify padding and normalization behaviour

younesbelkada committed:
ENH: Do not pass warning message in case `quantization_config` is in config but not passed as an arg (#28988)

* Update auto.py

* Update auto.py

* Update src/transformers/quantizers/auto.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

* Update src/transformers/quantizers/auto.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
ENH [`AutoQuantizer`]: enhance trainer + not supported quant methods (#28991)

* enhance trainer + not support quant methods

* remove all old logic

* add version

jon-tow committed:
Add `StableLM` (#28810)

* Add `StableLM`

* fix(model): re-create from `huggingface-cli add-new-model-like persimmon`

* fix: re-add changes to address comments

* fix(readme): add links to paper

* fix(tokenization_auto): remove `GPTNeoXTokenizerFastFast` ref

* fix(tests): re-add `@slow` decorator to integration tests

* fix(tests): import slow...

* fix(readme_hd): remove whitespace edit

* fix(tokenizer): auto tokenizer tuple

* skip doctests for `modeling_stablelm`

NielsRogge committed:
Add SiglipForImageClassification and CLIPForImageClassification (#28952)

* First draft

* Add CLIPForImageClassification

* Remove scripts

* Fix doctests

Commit:
AQLM quantizer support (#28928)

* aqlm init

* calibration and dtypes

* docs

* Readme update

* is_aqlm_available

* Simpler link in docs

* Test TODO real reference

* init _import_structure fix

* AqlmConfig autodoc

* integration aqlm

* integrations in tests

* docstring fix

* legacy typing

* Less typings

* More kernels information

* Performance -> Accuracy

* correct tests

* remoced multi-gpu test

* Update docs/source/en/quantization.md

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

* Update src/transformers/utils/quantization_config.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Brought back multi-gpu tests

* Update src/transformers/integrations/aqlm.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

* Update tests/quantization/aqlm_integration/test_aqlm.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

---------

Co-authored-by: Andrei Panferov <blacksamorez@yandex-team.ru>
Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>
Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>
Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

amyeroberts committed:
[`Doc`] Fix docbuilder - make `BackboneMixin` and `BackboneConfigMixin` importable from `utils`.  (#29002)

* Trigger doc build

* Test removing references

* Importable from utils

* Trigger another run on a new commit for testing

ji-huazhong committed:
Set the dataset format used by `test_trainer` to float32 (#28920)

Co-authored-by: unit_test <test@unit.com>

gante commented on setup.py:
```diff
@@ -258,6 +259,7 @@ def run(self):
 
 extras["tf"] = deps_list("tensorflow", "onnxconverter-common", "tf2onnx", "tensorflow-text", "keras-nlp")
 extras["tf-cpu"] = deps_list("tensorflow-cpu", "onnxconverter-common", "tf2onnx", "tensorflow-text", "keras-nlp")
+extras["pygtrie"] = deps_list("pygtrie")
 
 extras["torch"] = deps_list("torch", "accelerate")
```
```suggestion
extras["torch"] = deps_list("torch", "accelerate", "pygtrie")
```

Try this instead :) The `extras["pygtrie"]` you defined is not installed, as it is not included in `extras["all"]`

ahmed-moubtahij committed:
Update setup.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update setup.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

ahmed-moubtahij committed:
Update setup.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

muellerzr committed:
Introduce AcceleratorConfig dataclass (#28664)

* Introduce acceleratorconfig dataclass

* Extra second warn

* Move import

* Try moving import under is_accelerate_available

* Quality

* Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

* Clean

* Remove to_kwargs

* Change version

* Improve tests by including dispatch and split batches

* Improve reliability

* Update tests/trainer/test_trainer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

* Fixup tests and review nits

* Make tests pass

* protect import

* Protect import

* Empty-Commit

* Make training_args.to_dict handle the AcceleratorConfig

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

gante commented on tests/generation/test_utils.py:
```diff
@@ -3456,3 +3456,39 @@ def prepare_inputs_for_generation(self, *args, foo=False, encoder_outputs=None,
+
+@require_torch
+class TokenHealingTestCase(unittest.TestCase):
+    model_name_or_path = "TheBloke/deepseek-llm-7B-base-GPTQ"
+    tokenizer = AutoTokenizer.from_pretrained(model_name_or_path, use_fast=True)
+    completion_model = AutoModelForCausalLM.from_pretrained(
+        model_name_or_path,
+        device_map="auto",
+        trust_remote_code=False,
+        revision="main",
+        use_cache=True,
+    )
```
```suggestion
```

gante commented on tests/generation/test_utils.py:
```diff
@@ -3456,3 +3456,39 @@ def prepare_inputs_for_generation(self, *args, foo=False, encoder_outputs=None,
+            ("empty_prompt", "", ""),
+        ]
+    )
+    def test_prompts(self, name, input, expected):
```
```suggestion
    def test_prompts(self, name, input, expected):
        model_name_or_path = "TheBloke/deepseek-llm-7B-base-GPTQ"
        tokenizer = AutoTokenizer.from_pretrained(model_name_or_path, use_fast=True)
        completion_model = AutoModelForCausalLM.from_pretrained(
            model_name_or_path,
            device_map="auto",
            trust_remote_code=False,
            revision="main",
            use_cache=True,
        )
```

gante committed:
Update tests/generation/test_utils.py

gante committed:
Update tests/generation/test_utils.py

gante committed:
Update tests/generation/test_utils.py

gante committed:
Update tests/generation/test_utils.py

gante commented:
@Ayenem hope you don't mind, moving a few minor things around to make our CI happy 🤗

zucchini-nlp committed:
Fix flaky test vision encoder-decoder generate (#28923)

Ayenem commented:
> @Ayenem hope you don't mind, moving a few minor things around to make our CI happy 🤗

Thank you for taking the time!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28616). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gante commented on examples/flax/_tests_requirements.txt:
```diff
@@ -6,5 +6,4 @@ rouge-score
-torch
-accelerate
\ No newline at end of file
+torch
```
```suggestion
torch
accelerate
```

gante committed:
Update examples/flax/_tests_requirements.txt

gante committed:
Update examples/flax/_tests_requirements.txt

gante committed:
Update examples/flax/_tests_requirements.txt

gante committed:
Update setup.py

gante committed:
Update setup.py

gante committed:
Update setup.py

ahmed-moubtahij committed:
local variables instead of attributes

ahmed-moubtahij committed:
local variables instead of attributes

ahmed-moubtahij committed:
Merge branch 'token_healing' of github.com:Ayenem/transformers into token_healing

gante commented:
`examples_flax` is failing for reasons unrelated to this PR, I'm going to see if I can open a separate PR to fix it :)

gante commented:
Ok, actually, it has to be fixed in this PR: `pygtrie` has to be added to `/examples/flax/_tests_requirements.txt` :( 

As for the other failures, I recommend the following:
1. Rebase with `main`
2. run `make fixup` in the `transformers` folder
3. commit the resulting changes

If issues persist, I'll dive into the failures :)

ahmed-moubtahij committed:
add pygtrie

ahmed-moubtahij committed:
Merge branch 'token_healing' of github.com:Ayenem/transformers into token_healing

gante commented:
uhmmm something has gone wrong, the diff shouldn't be 18k lines long 😅

Ayenem commented:
> uhmmm something has gone wrong, the diff shouldn't be 18k lines long 😅

Indeed...I did a `git fetch upstream` followed by `git rebase upstream/main` then solved conflicts and pushed. I suppose I misinterpreted what you meant by rebasing with main 😓

gante commented:
@Ayenem That should be it: 
```
git remote add upstream https://github.com/huggingface/transformers.git
git fetch upstream
git rebase upstream/main
(handle merge conflicts)
git push origin your_branch -f
```

possibly something went wrong in the merge conflicts stage?

ahmed-moubtahij committed:
Revert "Merge branch 'token_healing' of github.com:Ayenem/transformers into token_healing"

This reverts commit b035d15be8c9aa2082f885558fc2b2dc91091f66, reversing
changes made to 8bc975520d5dcc93094ceb8de7057372c8bb432b.

ahmed-moubtahij committed:
Revert "Merge branch 'token_healing' of github.com:Ayenem/transformers into token_healing"

This reverts commit b035d15be8c9aa2082f885558fc2b2dc91091f66, reversing
changes made to 8bc975520d5dcc93094ceb8de7057372c8bb432b.

ahmed-moubtahij committed:
add pygtrie

ahmed-moubtahij committed:
add pygtrie

gante commented:
@Ayenem you're getting the best of CI problems, hehe

ahmed-moubtahij committed:
Update setup.py

ahmed-moubtahij committed:
Merge branch 'token_healing' of github.com:Ayenem/transformers into token_healing

Ayenem commented:
> @Ayenem you're getting the best of CI problems, hehe

Lucky me 😄 Rebase should've worked as expected now 🤞

gante commented:
@Ayenem looking at the test errors, other than sorting the merge conflicts (which should fix most of the problems), the new test needs `@require_auto_gptq`. Alternatively, a smaller model like `gpt2` can be used.

ahmed-moubtahij committed:
add require_auto_gptq

Ayenem commented:
> @Ayenem looking at the test errors, other than sorting the merge conflicts (which should fix most of the problems), the new test needs `@require_auto_gptq`. Alternatively, a smaller model like `gpt2` can be used.

I don't seem to have access to those conflicts locally and I don't have write access for resolving them from github 😕 
![image](https://github.com/huggingface/transformers/assets/50707385/6a210a82-48d8-4d61-b263-4c8268bd4b1c)

gante commented:
@Ayenem I have no more suggestions as well. When this happens, it often means something unexpected happened at rebase time. Most (or even all) CI errors we're seeing here are also already solved on `main` 😞 

May I suggest to:
1. Close this PR
2. Update `main` on your fork
3. Create a new branch on your fork, past these changes, and open a new PR?

Again, apologies for all this additional work 🤗
